File: 2501~3000/2785. Sort Vowels in a String/main.py
"""
Source: https://rb.gy/qj3uj
Date: 2023/7/23
Skill:
Ref:
Runtime: 167 ms, faster than 80.00%
Memory Usage: 19.8 MB, less than 100.00%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def sortVowels(self, s: str) -> str:
        vowels, vowel_set, string, res = [], set(), "AEIOUaeiou", ""
        for c in string:
            vowel_set.add(c)
        for c in s:
            if c in vowel_set:
                vowels.append(c)

        vowels.sort(key=lambda x: ord(x), reverse=True)
        for c in s:
            if c in vowel_set:
                res += vowels.pop()
            else:
                res += c

        return res


if __name__ == "__main__":
    s = Solution()

File: 2501~3000/2785. Sort Vowels in a String/main.cpp
/**
 * Source: t.ly/G2Mi2
 * Date: 2023/11/13
 * Skill:
 * Runtime: 20 ms, faster than 99.23% of C++ online submissions
 * Memory Usage: 10.96 MB, less than 99.38% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;

const int MX = 128;

class Solution {
public:
    string sortVowels(string s) {
        int freq[MX] = {0}, idx = 0;
        for (auto &c: s) {
            if (isVowel(c)) {
                freq[c]++;
            }
        }
        for (auto &c: s) {
            if (isVowel(c)) {
                while (freq[idx] == 0) idx++;
                c = idx;
                freq[idx]--;
            }
        }
        return s;
    }
    bool isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
                || c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
    }
};


int main() {
    Solution s;
}